derive/src/traits.rs

```diff
@@ -0,0 +1,209 @@
+      #assert_fields_are_pod
+    ))
+  }
+
+  fn check_attributes(attributes: &[Attribute]) -> Result<(), &'static str> {
+    let repr = get_repr(attributes);
+    match repr.as_ref().map(|repr| repr.as_str()) {
+      Some("C") => Ok(()),
+      Some("transparent") => Ok(()),
+      _ => {
+        Err("Pod requires the struct to be #[repr(C)] or #[repr(transparent)]")
+      }
+    }
+  }
+}
+
+pub struct Zeroable;
+
+impl Derivable for Zeroable {
+  fn ident() -> TokenStream {
+    quote!(::bytemuck::Zeroable)
+  }
+
+  fn struct_asserts(input: &DeriveInput) -> Result<TokenStream, &'static str> {
+    generate_fields_are_trait(input, Self::ident())
+  }
+}
+
+pub struct TransparentWrapper;
+
+impl TransparentWrapper {
+  fn get_wrapper_type(
+    attributes: &[Attribute], fields: &Fields,
+  ) -> Option<TokenStream> {
+    let transparent_param = get_simple_attr(attributes, "transparent");
+    transparent_param.map(|ident| ident.to_token_stream()).or_else(|| {
+      let mut types = get_field_types(fields);
+      let first_type = types.next();
+      if let Some(_) = types.next() {
+        // can't guess param type if there is more than one field
+        return None;
+      } else {
+        first_type.map(|ty| ty.to_token_stream())
+      }
+    })
+  }
+}
+
+impl Derivable for TransparentWrapper {
+  fn ident() -> TokenStream {
+    quote!(::bytemuck::TransparentWrapper)
+  }
+
+  fn generic_params(input: &DeriveInput) -> Result<TokenStream, &'static str> {
+    let fields = get_struct_fields(input)?;
+
+    Self::get_wrapper_type(&input.attrs, fields).map(|ty| quote!(<#ty>))
+      .ok_or("when deriving TransparentWrapper for a struct with more than one field you need to specify the transparent field using #[transparent(T)]")
+  }
+
+  fn struct_asserts(input: &DeriveInput) -> Result<TokenStream, &'static str> {
+    let fields = get_struct_fields(input)?;
+    let wrapped_type = match Self::get_wrapper_type(&input.attrs, fields) {
+      Some(wrapped_type) => wrapped_type.to_string(),
+      None => return Err(""), /* other code will already reject this derive */
+    };
+    let mut wrapped_fields = fields
+      .iter()
+      .filter(|field| field.ty.to_token_stream().to_string() == wrapped_type);
+    if let None = wrapped_fields.next() {
+      return Err("TransparentWrapper must have one field of the wrapped type");
+    };
+    if let Some(_) = wrapped_fields.next() {
+      Err("TransparentWrapper can only have one field of the wrapped type")
+    } else {
+      Ok(quote!())
+    }
+  }
+
+  fn check_attributes(attributes: &[Attribute]) -> Result<(), &'static str> {
+    let repr = get_repr(attributes);
+
+    match repr.as_ref().map(|repr| repr.as_str()) {
+      Some("transparent") => Ok(()),
+      _ => {
+        Err("TransparentWrapper requires the struct to be #[repr(transparent)]")
+      }
+    }
+  }
+}
+
+fn get_struct_fields<'a>(
+  input: &'a DeriveInput,
+) -> Result<&'a Fields, &'static str> {
```

Comment:
Does the lifetime `'a` need to be named in this function?

Reply:
since the &'static str has a named lifetime it's good to name the other lifetime, even if not absolutely required.